### PR TITLE
Fluent diff for new attributes

### PR DESCRIPTION
**Pull request by Pike**
This changelist is triggered by bug 1416451, where we don't show new FTL attributes.

I ended up refactoring the diff view, for which I refactored the test beforehand.

With the new setup, we can explicitly test the context data for the diff view for a particular content pair, without doing a ton of hg repo setup and such.

### CODE REVIEW

**Comment by Pike** [comment]
I haven't gone full monty on adjusting test_diff.py to the new test setup, I might do that as a follow-up. Either in this patch queue, or in a follow-up bug.

File: apps/pushes/views/diff.py [hunk 1]
```diff
@@ -29,6 +29,9 @@ class BadRevision(Exception):
 
 
 class DiffView(View):
+    moved = {}
+    copied = {}
```
Comment (stasm) [inline apps/pushes/views/diff.py]: How did this work before this addition? Is it necessary given that line 131 also initializes these dicts?

File: apps/pushes/views/diff.py [hunk 2]
```diff
@@ -172,104 +173,57 @@ def diffLines(self, path, action):
             p = getParser(path)
         except UserWarning:
             return None
-        if action == 'added':
-            a_entities = []
-            a_map = {}
-        else:
+        old_data = OrderedDict()
```
Comment (stasm) [inline apps/pushes/views/diff.py]: What kind of `data` is this? `diff_data`, `flat_data`, `ordered_entities`?

File: apps/pushes/views/diff.py [hunk 3]
```diff
@@ -172,104 +173,57 @@ def diffLines(self, path, action):
             p = getParser(path)
```
Comment (stasm) [inline apps/pushes/views/diff.py]: Please don't use 1-letter nor 2-letter variable names. Ever. `parser = .…`.

File: apps/pushes/views/diff.py [hunk 4]
```diff
@@ -172,104 +173,57 @@ def diffLines(self, path, action):
...
             content = self.content(path, self.rev2)
             try:
-                p.readContents(content)
-                c_entities, c_map = p.parse()
+                self.parse(p, content, new_data)
             except:
                 # consider doing something like:
                 # logging.warn('Unable to parse %s', path, exc_info=True)
                 return None
         ar = AddRemove()
```
Comment (stasm) [inline apps/pushes/views/diff.py]: `diff`, `differ` or even `add_remove` would all be better names for this instance.

File: apps/pushes/views/diff.py [hunk 5]
```diff
@@ -278,6 +232,15 @@ def content(self, path, rev):
         content = self._universal_newlines(content)
         return content
 
+    def parse(self, parser, content, data):
+        parser.readContents(content)
+        for entity in parser:
+            data[entity.key] = entity.val
+            if isinstance(entity, FluentEntity):
+                for fluent_attr in entity.attributes:
+                    data[u'{}\0{}'.format(entity.key, fluent_attr.key)] = \
```
Comment (stasm) [inline apps/pushes/views/diff.py]: Is the `\0` trick really necessary? Do I understand correctly that `data` is per file? If so, the period `.` is not legal as part of the identifier in Fluent files.

File: apps/pushes/views/diff.py [hunk 6]
```diff
@@ -278,6 +232,15 @@ def content(self, path, rev):
         content = self._universal_newlines(content)
         return content
 
+    def parse(self, parser, content, data):
```
Comment (stasm) [inline apps/pushes/views/diff.py]: Would it make sense to return an `OrderedDict` from this method? You could rename it to `parse_ordered` and use it like so:

```py
new_data.update(self.parse_ordered(parser, content))
```

File: apps/pushes/views/diff.py (hunk 1 above)
Comment (Pike) [reply apps/pushes/views/diff.py]: This wasn't used before, as I wasn't unit-testing this view at all, basically. All tests went through the initial .get(), which set these. Having these as dicts on the class isn't great, but it's not much of a problem. In practice, in each view, a new instance is instantiated, which overwrites the class ones. I also do that in tests, so the class-members ones are only ever read from, and in their empty state.

Alternatively I'd overwrite the constructor, but the default one is kind nice and generic, didn't want to touch that.

I'll add a comment.

File: apps/pushes/views/diff.py (hunk 2 above)
Comment (Pike) [reply apps/pushes/views/diff.py]: I think the difference between old and new is the relevant qualifier here. It maps from key-ish thing to value. To answer the question you didn't ask, if you need the variable name to be more specific, this needs a comment, not a variable name. I'll add a comment.

File: apps/pushes/views/diff.py (hunk 5 above)
Comment (Pike) [reply apps/pushes/views/diff.py]: We talked about this, and no, I'll remove the stunt.

File: apps/pushes/views/diff.py (hunk 1 above)
Comment (Pike) [reply apps/pushes/views/diff.py]: ... actually, trivial enough to subclass dict to not be mutable, so I did that.

File: apps/pushes/views/diff.py (hunk 2 above)
Comment (Pike) [reply apps/pushes/views/diff.py]: I added a comment to the top of the file to describe the algorithm.

We also talked about where to put the OrderedDict, and I went for the location where it was, but the .parse function now yields keys and values, so I can still call .update() on the result.

**Comment by Pike** [comment]
This is ready for review again.

The squashed patch isn't toooooo bad, but I'm leaning towards keeping this as a commit series, as some patches are quite substantial in how they change stuff.

I did test-coverage locally, and we're at a single line not covered by tests now for diff.py, which is the `raise RuntimeError` if hg status actually came back with a result we don't know. That's only happening if mercurial actually changes that API, so that's cool.

File: apps/pushes/views/diff.py (hunk 2 above)
Comment (stasm) [reply apps/pushes/views/diff.py]: > if you need the variable name to be more specific, this needs a comment, not a variable name. I'll add a comment.

This is a bad practice. Virtually every other name is better than `data`. How about `old_entities` and `new_entities`?

Comment (Pike) [reply apps/pushes/views/diff.py]: Well, the challenge is that this isn't something that shows up anywhere else. Entities would be wrong, as these are just values, and not even Entity.vals (due to FluentAttribute not being an Entity). `old_translations` vs `new_translations` could work?

Comment (stasm) [reply apps/pushes/views/diff.py]: Yeah, that's better too, thanks. Or perhaps `diffables`?
